src/util/StringUtils.cpp

```diff
@@ -78,3 +78,11 @@ auto StringUtils::iequals(const string& a, const string& b) -> bool {
 
     return result == 0;
 }
+
+auto StringUtils::isNumber(const string& input) -> bool {
+    for (char const& c: input) {
+        if (std::isdigit(c) == 0)
+            return false;
+    }
+    return true;
+}
```

Comment:
Check if it is better to use [`std::from_chars`](https://en.cppreference.com/w/cpp/utility/from_chars) here.
Especially if a conversion follows this function and when it's called in a loop. `std::from_chars` may the faster version.

Reply:
```cpp
                    } else if (StringUtils::isNumber(arg)) {
                        paletteIndex = std::stoi(arg);
                        namedColor = palette->getColorAt(paletteIndex);
```

This is how it's used in the `ToolMenuHandler` when the palette is read. I think here it is OK to trade speed for readability (unless someone has a lot of colors in the toolbar).

With the `std::from_chars` method the control flow would be more complicated as it's not as strict as I would like it to be.

I assume the could would need to look like the following:

```cpp
int tmp;
auto [p, ec]  = std::from_chars(arg.begin(), arg.end(), tmp);
if(p == arg.end()) paletteIndex = tmp;
...
```

Reply:
> I assume the could would need to look like the following:
> 
> ```c++
> int tmp;
> auto [p, ec]  = std::from_chars(arg.begin(), arg.end(), tmp);
> if(p == arg.end()) paletteIndex = tmp;
> ...
> ```

I am not exactly sure, but it looks good. I would probably check for the error, depending on the invariant at the end of the function call.
It also has the advantage, to catch overflow errors, and it works with the same locale ("C") as `std::atoi` / `std::stoi` If you want to use the latter you actually should check the errno flag. And this is not thread safe.

Reply:
I think I would rewrite it in this way then:

```cpp
// check for old color format of toolbar.ini
                    if (StringUtils::startsWith(arg, "0x")) {
                        namedColor = palette->getColorAt(colorIndex);
                        paletteIndex = namedColor.getIndex();

                        // set new COLOR Toolbar entry
                        std::ostringstream newColor("");
                        newColor << "COLOR(" << paletteIndex << ")";
                        dataItem->setName(newColor.str());
                        colorIndex++;
                    } else {
                        std::istringstream colorStream(arg);
                        colorStream >> paletteIndex;
                        if (!colorStream.eof() || colorStream.fail()) {
                            g_warning("Toolbar:COLOR(...) has to start with 0x, get color: %s", arg.c_str());
                            continue;
                        }
                        namedColor = palette->getColorAt(paletteIndex);
                    }
```

I would still have two questions for my understanding:
1. Why would I need to check errno? Wouldn't it simply through an exception if there were an `std::out_of_range` error?
2. Why would thread safety be of concern here?